apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelperImpl.java

```diff
@@ -86,4 +109,49 @@ public Span startJmsProducerSpan(Destination destination, Message message) {
         }
         return span;
     }
+
+    @VisibleForAdvice
+    @Override
+    @Nullable
+    public MessageListener wrapLambda(@Nullable MessageListener listener) {
+        // the name check also prevents from wrapping twice
+        if (listener != null && listener.getClass().getName().contains("/")) {
+            return messageListenerWrapperObjectPool.createInstance().wrap(listener);
+        }
+        return listener;
+    }
+
+    private void recycle(MessageListenerWrapper messageListenerWrapper) {
+        messageListenerWrapperObjectPool.recycle(messageListenerWrapper);
+    }
+
+    public class MessageListenerWrapper implements MessageListener, Recyclable {
+
+        @Nullable
+        private MessageListener delegate;
```

Comment:
does this need to be volatile?

Reply:
It better be!

Reply:
Now that we don't pool them- irrelevant. Will just make final and initialize on construction.